torch/csrc/jit/passes/annotate_effects.cpp

```diff
@@ -0,0 +1,227 @@
+#include "torch/csrc/jit/passes/annotate_effects.h"
+
+#include <set>
+#include "torch/csrc/jit/passes/dead_code_elimination.h"
+
+namespace torch {
+namespace jit {
+namespace {
+
+/**
+ * AnnotateEffects
+ *
+ * This pass annotates effectful operations (such as ones that mutate existing
+ * values) to prevent subsequent passes from re-ordering ops in a way that
+ * changes the meaning of the program.
+ *
+ * It does this by threading a "world token" value through nodes that use
+ * mutable values. This models effects explicitly in the IR and forces all
+ * annotated nodes to be linearized during optimization.
+ *
+ * The world token is threaded directly through any nodes that mutate values.
+ * For purely functional operators, their node will be "fenced" by two
+ * `prim::MemoryFence` nodes that take world tokens as their input.
+ */
+class AnnotateEffectsImpl {
+ public:
+  void annotateEffects(Graph* g) {
+    // TODO(suo): We need to change this so that the world token is an input
+    // and output of the graph. That would require changing a bunch of interop
+    // and batching code, so leaving it out for now.
+    //
+    // auto curToken = g->addInput("world")->setType(WorldType::get());
```

Comment:
Discussed in person: we can add `g->entryWorld()` and `g->exitWorld()` to the graph. These would not be members of `inputs()` or `outputs()` of the graph but come from separate "special" nodes (which, similar to prim::Params, are not in the node list) that are used only to capture effects.